File: oop1/Seeker.cpp
#include "Seeker.h"

#include <cstdlib>
#include <iostream>

using std::cout;
using std::endl;

Seeker::Seeker(const std::string& name, int level, int startEnergy)
    : MeleeCharacter(name, level, 30 + level * 3, 50 + level * 4),
      energy(startEnergy),
      dodgeChance(20 + agility / 5),
      isStealthed(false),
      stealthSeconds(0.0f) {
    cout << "Создан Искатель " << getName() << " уровень: " << getLevel() << endl;
    cout << "  Способности: Теневой клинок, Скрытность" << endl;
}

Seeker::Seeker(const std::string& name, int level)
    : Seeker(name, level, 100) {}

Seeker::Seeker(const std::string& name)
    : Seeker(name, 1, 80) {}

void Seeker::attack(Character& target) {
    int damage = calculateDamage();
    if (isStealthed) {
        damage *= 2;
        isStealthed = false;
        stealthSeconds = 0.0f;
        cout << getName() << " атакует из скрытности! ";
    } else {
        cout << getName() << " наносит быстрый двойной удар! ";
    }

    cout << "Урон: " << damage << endl;
    target.takeDamage(damage);
    energy += 10;
}

void Seeker::useAbility() {
    if (energy >= 30) {
        cout << getName() << " использует 'Теневой клинок'!" << endl;
        dodgeChance += 15;
        energy -= 30;
        cout << "  Шанс уворота увеличен до " << dodgeChance << "%" << endl;
    } else {
        cout << getName() << ": Недостаточно энергии!" << endl;
    }
}

void Seeker::takeDamage(int damage) {
    if (tryDodge()) return;
    Character::takeDamage(damage);
}

void Seeker::useSpecial() {
    enterStealth();
}

void Seeker::enterStealth() {
    if (energy >= 20) {
        isStealthed = true;
        stealthSeconds = 10.0f;
        energy -= 20;
        dodgeChance += 25;
        cout << getName() << " скрывается в тенях..." << endl;
        cout << "  Шанс уворота: " << dodgeChance << "%" << endl;
    }
}

bool Seeker::tryDodge() const {
    int roll = rand() % 100;
    bool dodged = (roll < dodgeChance);
    if (dodged) {
        cout << getName() << " уворачивается от атаки!" << endl;
    }
    return dodged;
}

void Seeker::onUpdate(float deltaSeconds) {
    if (!isStealthed) return;

    stealthSeconds -= deltaSeconds;
    if (stealthSeconds <= 0.0f) {
        isStealthed = false;
        stealthSeconds = 0.0f;
        cout << getName() << " выходит из скрытности." << endl;
    }
}

void Seeker::showSeekerStats() const {
    showStats();
    cout << "Энергия: " << energy << "/100" << endl;
    cout << "Шанс уворота: " << dodgeChance << "%" << endl;
    cout << "Состояние: " << (isStealthed ? "В скрытности" : "Видим") << endl;
}

void Seeker::printInfo() const {
    cout << name << " ";
    Character::printInfo();
    cout << "  Энергия: " << energy
         << ", шанс уворота: " << dodgeChance << "%" << endl;
}

int Seeker::getEnergy() const { return energy; }
int Seeker::getDodgeChance() const { return dodgeChance; }
bool Seeker::getStealthState() const { return isStealthed; }

void Seeker::setEnergy(int e) {
    if (e >= 0 && e <= 100) {
        energy = e;
    }
}

File: oop3/src/main/java/com/game/entity/Seeker.java
package com.game.entity;

import jakarta.persistence.Column;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import jakarta.persistence.PrimaryKeyJoinColumn;
import jakarta.persistence.Table;

import lombok.Getter;
import lombok.Setter;

@Entity
@Table(name = "seeker")
@PrimaryKeyJoinColumn(name = "id", referencedColumnName = "id")
@DiscriminatorValue("Seeker")
@Getter
@Setter
public class Seeker extends MeleeCharacter {

    @Column(nullable = false)
    private Integer energy;

    @Column(name = "dodge_chance", nullable = false)
    private Integer dodgeChance;

    @Column(name = "is_stealthed", nullable = false)
    private Boolean isStealthed;

    @Column(name = "stealth_seconds", nullable = false)
    private Float stealthSeconds;

    public Seeker() {
        this.energy = 50;
        this.dodgeChance = 15;
        this.isStealthed = false;
        this.stealthSeconds = 3.0f;
    }

    public Seeker(Long id, String name, Integer level, Integer health,
      Integer maxHealth, Integer baseDamage, Integer regenAmount, Float regenIntervalSec,
      Integer strength, Integer agility, Integer energy, Integer dodgeChance,
      Boolean isStealthed, Float stealthSeconds) {
        super(id, name, level, health, maxHealth, baseDamage, regenAmount, regenIntervalSec,
          strength, agility);
        this.energy = energy;
        this.dodgeChance = dodgeChance;
        this.isStealthed = isStealthed;
        this.stealthSeconds = stealthSeconds;
    }

    public void attack(Character target) {
        int damage = calculateDamage();
        if (isStealthed) {
            damage *= 2;
            isStealthed = false;
            stealthSeconds = 0.0f;
            System.out.println(getName() + " атакует из скрытности! ");
        } else {
            System.out.println(getName() + " наносит быстрый двойной удар! ");
        }

        System.out.println("Урон: " + damage);
        target.takeDamage(damage);
        energy += 10;
    }

    public void useAbility() {
        if (energy >= 5) {
            System.out.println(getName() + " использует 'Теневой клинок'!");
            dodgeChance += 15;
            energy -= 5;
            System.out.println("  Шанс уворота увеличен до " + dodgeChance + "%");
        } else {
            System.out.println(getName() + ": Недостаточно энергии!");
        }
    }

    public void takeDamage(int damage) {
        if (tryDodge()) return;
        super.takeDamage(damage);
    }

    public void useSpecial() {
        enterStealth();
    }

    public void enterStealth() {
        if (energy >= 5) {
            isStealthed = true;
            stealthSeconds = 10.0f;
            energy -= 5;
            dodgeChance += 20;
            System.out.println(getName() + " скрывается в тенях...");
            System.out.println("  Шанс уворота: " + dodgeChance + "%");
        }
    }

    public boolean tryDodge() {
        int roll = (int) (Math.random() * 100);
        boolean dodged = roll < dodgeChance;
        if (dodged) {
            System.out.println(getName() + " уворачивается от атаки!");
        }
        return dodged;
    }

    public void onUpdate(float deltaSeconds) {
        if (!isStealthed) return;

        stealthSeconds -= deltaSeconds;
        if (stealthSeconds <= 0.0f) {
            isStealthed = false;
            stealthSeconds = 0.0f;
            System.out.println(getName() + " выходит из скрытности.");
        }
    }

    public void showSeekerStats() {
        showStats();
        System.out.println("Энергия: " + energy + "/100");
        System.out.println("Шанс уворота: " + dodgeChance + "%");
        System.out.println("Состояние: " + (isStealthed ? "В скрытности" : "Видим"));
    }

    public void printInfo() {
        System.out.println(getName() + ": энергия " + energy + ", шанс уворота " + dodgeChance + "%");
    }

    public void setEnergy(Integer energy) {
        if (energy == null) {
            this.energy = 50;
            return;
        }
        this.energy = Math.max(0, Math.min(energy, 1000));
    }
}
